Improves Warning messaging

tfishersp opened the pull request:
Improves on the original warning.warn message (for when an envfile wasn't found) by including the search path used for a given file.

Commit:
Improves Warning messaging

Improves on the original warning.warn message (for when an envfile wasn't
found) by including the search path used for a given file.

rconradharris commented on envparse.py:
```diff
@@ -114,6 +114,22 @@ def __call__(self, var, default=NOTSET, cast=None, subcast=None,
         return value
 
     @classmethod
+    def track_path(cls, path="", tracking_table={}):
```
The use of a mutable type as default value is a little dangerous. (see http://docs.python-guide.org/en/latest/writing/gotchas/). Recommend using `None` as default, and setting to `{}` inside function.

rconradharris commented on envparse.py:
```diff
@@ -183,6 +199,8 @@ def read_envfile(path=None, **overrides):
                 content = f.read()
         except getattr(__builtins__, 'FileNotFoundError', IOError):
+            # Append the path we tried (and failed) to find to a tracking table
+            traversed_paths = Env.track_path(path)
```
Nit, but Env can be `cls` here.

tfishersp replied:
Where it's used inside a static method, so there's no `cls` binding param AFAIK.

rconradharris commented on envparse.py:
```diff
@@ -114,6 +114,22 @@ def __call__(self, var, default=NOTSET, cast=None, subcast=None,
         return value
 
     @classmethod
```
This probably doesn't need to be a separate function. Since it's not likely to be used elsewhere, just adding a dict to `read_envfile` would be simpler.

tfishersp replied:
I agree it's unlikely to be used elsewhere (and simply tracking it in the function was my first thought), but the `read_envfile` method [calls itself] (https://github.com/rconradharris/envparse/blob/master/envparse.py#L191-L192), so I used a classmethod to "log-sideways" as otherwise it would overwrite a var placed in the body.  The class method approach had a nifty little side effect of tracking the lookup attempts on each class, so I went that way.

The other way I thought to do this would be passing the "`traversed_paths`" along in the `read_envfile()` params.  I'l hack that up quickly for this PR.

rconradharris commented:
Looks useful.

Think an additional function for this is probably overkill, just adding a `list` of paths we've traversed to `read_envfile` should be good enough.

Commit:
Removes function in lieu of simple list

Instead of implementing the (missing path) logging on the class level,
 instead just track as a list in the read_envfile params.

ghost commented:
Thanks for saying that it looks useful and for quickly responding to the PR (👍 ).

I think the last commit is what you had in mind, but let me know if I misunderstood and I can modify it.

ghost commented:
It is worth noting that the change to only use a list in that fashion means that the `UserWarning` will contain all failed path lookups (i.e. `read_envfile('a'); read_envfile('b')` will show warnings on 'b' that contain paths from 'a').  Let me know if you want me to write in the "only show warnings for `filename`" behavior.

MichaelCurrie commented:
@rconradharris are you still maintaining this repo?  The last commit is 7 years ago.  Thanks
